// testcontainers/dist/src/modules/elasticsearch/elasticsearch-container.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartedElasticsearchContainer = exports.ElasticsearchContainer = void 0;
const __1 = require("../..");
const abstract_started_container_1 = require("../abstract-started-container");
const ELASTIC_SEARCH_HTTP_PORT = 9200;
class ElasticsearchContainer extends __1.GenericContainer {
    constructor(image = "elasticsearch:7.17.7") {
        super(image);
    }
    start() {
        const _super = Object.create(null, {
            start: { get: () => super.start }
        });
        return __awaiter(this, void 0, void 0, function* () {
            this.withExposedPorts(...(this.hasExposedPorts ? this.opts.exposedPorts : [ELASTIC_SEARCH_HTTP_PORT]))
                .withEnvironment({ "discovery.type": "single-node" })
                .withCopyContentToContainer([
                {
                    content: "-Xmx2G\n",
                    target: "/usr/share/elasticsearch/config/jvm.options.d/elasticsearch-default-memory-vm.options",
                },
            ])
                .withStartupTimeout(120000);
            return new StartedElasticsearchContainer(yield _super.start.call(this));
        });
    }
}
exports.ElasticsearchContainer = ElasticsearchContainer;
class StartedElasticsearchContainer extends abstract_started_container_1.AbstractStartedContainer {
    constructor(startedTestContainer) {
        super(startedTestContainer);
        this.startedTestContainer = startedTestContainer;
        this.httpPort = this.getMappedPort(ELASTIC_SEARCH_HTTP_PORT);
    }
    getHttpUrl() {
        return `http://${this.getHost()}:${this.httpPort}`;
    }
}
exports.StartedElasticsearchContainer = StartedElasticsearchContainer;
