// testcontainers/dist/src/docker/client/strategy/unix-socket-strategy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnixSocketStrategy = void 0;
const dockerode_1 = __importDefault(require("dockerode"));
const fs_1 = require("fs");
class UnixSocketStrategy {
    getDockerClient() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                uri: "unix:///var/run/docker.sock",
                dockerode: new dockerode_1.default({ socketPath: "/var/run/docker.sock" }),
                composeEnvironment: {},
                allowUserOverrides: true,
            };
        });
    }
    isApplicable() {
        return (process.platform === "linux" || process.platform === "darwin") && (0, fs_1.existsSync)("/var/run/docker.sock");
    }
    getName() {
        return "UnixSocketStrategy";
    }
}
exports.UnixSocketStrategy = UnixSocketStrategy;
